src/mame/machine/hp2640_tape.cpp

```diff
@@ -0,0 +1,547 @@
+		} else if (BIT(m_cmd_reg , CMD_REG_WRITE_BIT) && BIT(m_cmd_reg , CMD_REG_GAP_WRITE_BIT)) {
+			m_current_op = hp_dc100_tape_device::OP_ERASE;
+		} else {
+			m_current_op = hp_dc100_tape_device::OP_IDLE;
+		}
+
+		if (!set_speed()) {
+			start_rd_wr();
+		}
+	}
+}
+
+uint8_t hp2640_tape_device::status_r()
+{
+	uint8_t res = 0;
+
+	if (m_tach_latch) {
+		BIT_SET(res , 7);
+	}
+	if (m_byte_ready) {
+		BIT_SET(res , 6);
+	}
+	if (m_gap) {
+		BIT_SET(res , 5);
+	}
+	if (m_hole_latch) {
+		BIT_SET(res , 4);
+	}
+	if (m_tach_div2) {
+		BIT_SET(res , 3);
+	}
+	if (m_rip) {
+		BIT_SET(res , 2);
+	}
+	if (!m_drives[ 1 ]->cart_out_r()) {
+		BIT_SET(res , 1);
+	}
+	if (!m_drives[ 0 ]->cart_out_r()) {
+		BIT_SET(res , 0);
+	}
+
+	m_tach_latch = false;
+	m_hole_latch = false;
+	update_irq();
+
+	LOG("STS=%02x\n" , res);
+	return res;
+}
+
+void hp2640_tape_device::data_w(uint8_t data)
+{
+	LOG("DATA W=%02x\n" , data);
+	m_data_rd = data;
+	m_byte_ready = false;
+	update_irq();
+}
+
+uint8_t hp2640_tape_device::data_r()
+{
+	m_byte_ready = false;
+	update_irq();
+
+	LOG("DATA R=%02x\n" , m_data_rd);
+	return m_data_rd;
+}
+
+uint8_t hp2640_tape_device::poll_r() const
+{
+	return m_irq ? 0x80 : 0x00;
+}
+
+void hp2640_tape_device::device_add_mconfig(machine_config &config)
+{
+	for (unsigned i = 0; i < 2; i++) {
+		auto& finder = m_drives[ i ];
+
+		HP_DC100_TAPE(config , finder , 0);
+		// Acceleration: 2000 in/s^2
+		finder->set_acceleration(ACCELERATION);
+		// Slow speed: 10 ips, Fast speed: 60 ips
+		finder->set_set_points(SLOW_SPEED , FAST_SPEED);
+		// 58.4 ticks per inch
+		finder->set_tick_size(TACH_TICK_LENGTH);
+		// Manchester encoded data
+		finder->set_image_format(hti_format_t::HTI_MANCHESTER_MOD);
+		// Moving when speed is >1 ips
+		finder->set_go_threshold(MOVING_THRESHOLD);
+
+		finder->hole().set([this , i](int state) { hole_w(i , state); });
+		finder->tacho_tick().set([this , i](int state) { tacho_tick_w(i , state); });
+		finder->motion_event().set([this , i](int state) { motion_w(i , state); });
+		finder->rd_bit().set([this , i](int state) { rd_bit_w(i , state); });
+		finder->wr_bit().set([this , i](int state) { return wr_bit_r(i); });
+	}
+}
+
+void hp2640_tape_device::device_start()
+{
+	m_irq_handler.resolve_safe();
+	m_led0_handler.resolve_safe();
+	m_led1_handler.resolve_safe();
+
+	m_gap_timer = timer_alloc(GAP_TMR_ID);
+	m_cell_timer = timer_alloc(CELL_TMR_ID);
+}
+
+void hp2640_tape_device::device_reset()
+{
+	m_selected_drive = 0;
+
+	// All bits set to 1 but 3 & 0
+	m_cmd_reg = ~0;
+	BIT_CLR(m_cmd_reg , CMD_REG_WRITE_BIT);
+	BIT_CLR(m_cmd_reg , CMD_REG_RUN_BIT);
+
+	m_modulus = MODULUS_RESET;
+	m_cell_cnt = 6;
+	m_bit_cnt = 8;
+	m_tach_div2 = false;
+	m_tach_latch = false;
+	m_hole_latch = false;
+	m_byte_ready = false;
+	m_irq = true;
+	m_bit_sync = false;
+	m_isf = false;
+	m_gap = true;
+	m_prev_gap = true;
+	m_rip = false;
+	m_current_op = hp_dc100_tape_device::OP_READ;
```

Comment:
At least some of these should probably be registered for save states.

Reply:
see device_start